core/src/replay_stage.rs

```diff
@@ -827,18 +899,51 @@ impl ReplayStage {
             descendants,
         );
 
-        for d in slot_descendants
-            .iter()
-            .chain(std::iter::once(&duplicate_slot))
-        {
-            // Clear the progress map of these forks
-            let _ = progress.remove(d);
+        // Grab the Slot and BankId's of the banks we need to purge, then clear the banks
+        // from BankForks
+        let (slots_to_purge, removed_banks): (Vec<(Slot, BankId)>, Vec<Arc<Bank>>) = {
```

Comment:
After purging from `BankForks` could we just get new `ancestors` and `descendants` rather than trying to keep them consistent above with `purge_ancestors_descendants`? I think that would be safer and could avoid future bugs

Reply:
Yeah I thought about this as well, but was concerned that if we had a large tree with a bunch of slots on the same branch being deleted, it would cause a bunch of repetitive `n^2` operations to regenerate the ancestors and descendants.

For instance if you had a branch

```
1 -> 2 -> 3 -> 4 -> 5
   | -------BIG BRANCH
```

and 2, 3 ,4, 5 were all being dumped, then you would either:

1) Regenerate the ancestors/descendants after every single slot was dumped.
2) Only regenerate the ancestors/descendants at the end when all the dumping was done. But this means you have an inconsistency between the data structures in earlier vs later slots, i.e. `BankForks/AccountsDb/Blockstore` could already be missing some of the slots that had been cleared earlier, so you would have to handle those cases instead of hard asserting that they're consistent as we do right now.

Reply:
Got it, yeah this is more efficient by not recreating ancestors, but at least descendants is updated with each removal from bank forks. Not sure if the efficiency savings are worth it if it means a bug is introduced down the road. Where is the consistency check you're referring to?

Reply:
> Only regenerate the ancestors/descendants at the end when all the dumping was done. But this means you have an inconsistency between the data structures in earlier vs later slots, i.e. `BankForks/AccountsDb/Blockstore` could already be missing some of the slots that had been cleared earlier, so you would have to handle those cases instead of hard asserting that they're consistent as we do right now.

I meant if you followed this then in places like here: https://github.com/solana-labs/solana/blob/b69afd9d88a1a13de6437d6e54bd974f3728e106/core/src/replay_stage.rs#L946-L954.  If you 

1) Cleared a later descendant and didn't update the  the `ancestors/descendants`  
2) Then cleared an ancestor, the `slot_descendants` of that ancestor will not be updated and the assertion panics

Reply:
> Got it, yeah this is more efficient by not recreating ancestors, but at least descendants is updated with each removal from bank forks. Not sure if the efficiency savings are worth it if it means a bug is introduced down the road. Where is the consistency check you're referring to?

It's an n^2 operation to recreate all the ancestors and descendants. If there's a large number of banks, let's say 500 from some duplicate forking, that could potentially be a lot of scans

Reply:
https://github.com/solana-labs/solana/pull/18160/commits/1dac1a8933029b0bb628a5c4e3de4333a5c4ca83, seems to be taking multiple seconds